### DISCUSSION TITLE
[General] Cannot use 2 or more parameters in the paymaster's innerInput data

### DISCUSSION THREAD

**Discussion by lutr0** [zkSync-Community-Hub/zksync-developers#148] [General] [answered]
### Environment

Testnet

### Provide a brief description of the functionality you're trying to implement.

I am trying to deploy a custom paymaster implementation that verify a second signature by passing data to the innerInput.

### What is the specific issue or error you're encountering?

I can pass one parameter (e.g. the expiration timestamp), but I receive the error below if I try to add more

### Can you share the error messages or logs you're receiving, if any?

```
1 failing

  1) ERC20Paymaster
       Should validate and pay for paymaster transaction:
     Error: processing response error (body="{\"jsonrpc\":\"2.0\",\"error\":{\"code\":3,\"message\":\"Execution error: Failed to execute next transaction: Paymaster validation error: Error function_selector = 0x, data = 0x\",\"data\":\"0xf92b35b5d25a58b78e93049a6d71aa6917bb24c2c38062a5b9ed55b1d26b34ed\"},\"id\":64}\n", error={"code":3,"data":"0xf92b35b5d25a58b78e93049a6d71aa6917bb24c2c38062a5b9ed55b1d26b34ed"}, requestBody="{\"method\":\"eth_sendRawTransaction\",\"params\":[\"0x71f901ea8080840ee6b280835b8d809403513c07a7885415c025d29052795df24a9ea0c780b84440c10f1900000000000000000000000029a2d449ba9d67a49a05cd6f24f2bdfc167bbfd2000000000000000000000000000000000000000000000000000000000000000582010480808201049429a2d449ba9d67a49a05cd6f24f2bdfc167bbfd282c350c0b841ddb93500356ef10753ad904762b3e48d71908bd0c5e635f78961bcbed99378274d609e990149e8b9ef387cddd6026f3cda68667dadaf060da2ce660c1885a2561bf9011c942cebee2da07504bf9834a8306a1c34a60de8c391b90104949431dc00000000000000000000000003513c07a7885415c025d29052795df24a9ea0c70000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000006000000000000000000000000000000000000000000000000000000000000000610000000000000000000000000000000000000000000000000000000000000b1eb7e7dcefe60005b888aff952596e6e34dc7d6c4b686b0c8f3052393dead7fc2f24de593ba1d7d4b0f9e5989a5fad708af0586409a80936472a435119bec5ce281b00000000000000000000000000000000000000000000000000000000000000\"],\"id\":64,\"jsonrpc\":\"2.0\"}", requestMethod="POST", url="http://localhost:8011", code=SERVER_ERROR, version=web/5.7.1)
      at Logger.makeError (node_modules/@ethersproject/logger/src.ts/index.ts:269:28)
      at Logger.throwError (node_modules/@ethersproject/logger/src.ts/index.ts:281:20)
      at /home/joe/Dev/0Ondefy/ondefy-paymaster/node_modules/@ethersproject/web/src.ts/index.ts:341:28
      at step (node_modules/@ethersproject/web/lib/index.js:33:23)
      at Object.next (node_modules/@ethersproject/web/lib/index.js:14:53)
      at fulfilled (node_modules/@ethersproject/web/lib/index.js:5:58)
      at processTicksAndRejections (node:internal/process/task_queues:95:5)



error Command failed with exit code 1.
```

### Have you made any recent changes to the contract before encountering this issue?

N/A

### Are there any external libraries or contracts that your contract interacts with?

No

### Test file
```ts
const innerInput = ethers.utils.solidityPack(["uint256", "bytes"], [expiration, SignedMessageHash]);

    const paymasterParams = utils.getPaymasterParams(
      paymaster.address.toString(),
      {
        type: payType,
        token: token,
        minimalAllowance,
        innerInput: innerInput,
      }
    );

    console.log("Paymaster params: ", paymasterParams);

    await (
      await erc20.connect(userWallet).mint(user.address, 5, {
        maxPriorityFeePerGas: ethers.BigNumber.from(0),
        maxFeePerGas: gasPrice,
        gasLimit: GAS_LIMIT,
        customData: {
          paymasterParams: paymasterParams,
          gasPerPubdata: utils.DEFAULT_GAS_PER_PUBDATA_LIMIT,
        },
      })
    ).wait();
  }
  ```
### Solidity
  
  ```sol
if (paymasterInputSelector == IPaymasterFlow.approvalBased.selector) {
            // While the transaction data consists of address, uint256 and bytes data,
            // the data is not needed for this paymaster
            (address token, uint256 amount, bytes memory data) = abi.decode(
                _transaction.paymasterInput[4:],
                (address, uint256, bytes)
            );

            // Unwrap the data
            (uint256 expiration, bytes memory signedMessage) = abi.decode(
                data,
                (uint256, bytes)
            );
```

### Have you tried to isolate the problem, and if so, what were the results?

I tried using different parameters, but I always get an error with 2+ parameters.
It works with only one

### What steps have you already taken to try to resolve the issue?

I am running the in-memory node

**Comment by dutterbutter** [comment]
Thanks for posting!! Let me reproduce on my side and I will follow up.

**Reply by lutr0** [reply]
Hi @dutterbutter, I wonder if this one may also be solved by using arrayifi as in #156.
Still didn't manage to make it work on my end

**Reply by dutterbutter** [reply]
I havent had a chance to debug this further yet, but I plan to this week. My apologies on the delay. We want to make this apart of the [paymaster-examples](https://github.com/matter-labs/paymaster-examples) repo as well so others can see a working example.

**Reply by dutterbutter** [reply, marked as answer]
Hey @lutr0 - thanks for posting these great questions. Please keep them coming 💪 

I was able to pass in two params using the following example: 

```
const value1 = ethers.utils.hexlify(ethers.utils.toUtf8Bytes("dutterbutter"));
const value2 = ethers.utils.hexlify(ethers.utils.toUtf8Bytes("zk_dutterbutter"));

const paymasterParams = utils.getPaymasterParams(
  paymaster.address.toString(),
  {
    type: payType,
    token: token,
    minimalAllowance,
    innerInput: ethers.utils.defaultAbiCoder.encode(
      ["bytes", "bytes"],
      [ethers.utils.arrayify(value1), ethers.utils.arrayify(value2)]
  ),
  }
);
```

And then on the contract side:

```
function bytesToString(bytes memory data) internal pure returns (string memory) {
    return string(data);
}

if (paymasterInputSelector == IPaymasterFlow.approvalBased.selector) {
      // While the transaction data consists of address, uint256 and bytes data,
      // the data is not needed for this paymaster
      (address token, uint256 amount, bytes memory data) = abi.decode(
          _transaction.paymasterInput[4:],
          (address, uint256, bytes)
      );

      (string memory strValue1, string memory strValue2) = abi.decode(data, (string, string));
      console.log("value1: ", bytesToString(bytes(strValue1)));
      console.log("value2: ", bytesToString(bytes(strValue2)));
```

Logs from era-test-node 👇 
<img width="587" alt="Screenshot 2023-10-03 at 8 37 38 AM" src="https://github.com/zkSync-Community-Hub/zkync-developers/assets/29983536/eb86eafe-6ddf-447e-bc0b-8d179339ca6c">


**Note:** There is likely optimizations to be done here but figured I would share 💯

**Reply by lutr0** [reply]
Thank you so much, it's really helpful!
How would it change if a value were a `uint256` or an `address`?

**Comment by inesarmetessa** [comment]
It seems like there's an issue with passing multiple parameters to the paymasterParams in your test file. Ensure that the encoding and decoding of the parameters are correctly implemented both in the test file and in the Solidity contract. Double-check the data types and the ABI encoding to make sure they match on both ends. Additionally, carefully review the error message to identify any specific details that could provide insights into the root cause of the problem.

**Comment by dawn71v** [comment]
The issue seems to arise when passing multiple parameters to the innerInput of the paymasterParams. Ensure that the encoding and decoding of the parameters are correctly implemented on both the JavaScript and Solidity sides to handle multiple parameters.

**Comment by dunminecom** [comment]
It seems like there might be an issue with how the parameters are being encoded or decoded, particularly when passing multiple parameters. Double-check the encoding and decoding process to ensure that it aligns with the expected format on both ends. Additionally, consider testing with different parameter combinations to pinpoint the exact source of the error.
